File: 백준/Silver/10451. 순열 사이클/순열 사이클.py
import sys
input = sys.stdin.readline

T = int(input())

for _ in range(T) :
    N = int(input())
    num = [0] + list(map(int, input().split()))
    v = [False for _ in range(N+1)]
    cnt = 0
    for i in range(1, N+1) :
        if v[i] :
            continue
        a = num[i]
        cnt+=1
        while a != i :
            v[a] = True
            a = num[a]
    print(cnt)

File: 백준/Silver/10451. 순열 사이클/순열 사이클.java
import java.io.*;
import java.util.*;

public class Main {
	static int N;
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st;

		int T = Integer.parseInt(br.readLine());
		
		for (int TC = 0; TC < T; TC++) {
			N = Integer.parseInt(br.readLine());
			
			int cnt = 0;
			
			int[] num = new int[N+1];
			st = new StringTokenizer(br.readLine());
			for (int i = 1; i <= N; i++) {
				num[i] = Integer.parseInt(st.nextToken());
			}
			
			boolean[] v = new boolean[N+1];
			for (int i = 1; i <= N; i++) {
				if (v[i]) continue;
				cnt++;
				int a = i;
				while (!v[a]) {
					v[a] = true;
					a = num[a];
				}
			}
			sb.append(cnt).append("\n");
		}
		System.out.println(sb);
	}
}